### PR TITLE
feat: detect same commit on project and sync branch

### PR DESCRIPTION
**Pull request by ysndr**
`ManagedEnvironment::pull` will return `PullResult::UpToDate`
if both the project specific branch and it's upstream counterpart
are pointing at the same revision.

### CODE REVIEW

File: cli/flox-rust-sdk/src/models/environment/managed_environment.rs [hunk 1]
```diff
@@ -1073,6 +1080,7 @@ impl ManagedEnvironment {
 
     pub fn pull(&mut self, force: bool) -> Result<(), ManagedEnvironmentError> {
```
Comment (ysndr) [inline cli/flox-rust-sdk/src/models/environment/managed_environment.rs:1081]: Currently I return an error variant if the environment is already up-to-date and match on that.
Alternatively I can change this signature to `Result<PullState, ManagedEnvironmentError>` where PullState is

```
enum PullState {
	Updated
	UpToDate
}
```

(or a bool, but that doesn't read as clearly IMO).

Comment (ysndr) [reply cli/flox-rust-sdk/src/models/environment/managed_environment.rs]: The error turned out to interact weirdly with `push`.

File: cli/flox/src/commands/environment.rs [hunk 2]
```diff
@@ -1653,13 +1661,16 @@ impl Pull {
...
+    ) -> Result<PullResult, EnvironmentError2> {
+        let mut env = ManagedEnvironment::open(flox, pointer, dot_flox_path)?;
+        match env.pull(force)? {
+            state @ PullResult::UpToDate => Ok(state),
+            // only build if the environment was updated
+            state @ PullResult::Updated => {
+                env.build(flox)?;
+                Ok(state)
+            },
+        }
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:1673]: nit/slight preference
```suggestion
        let state = env.pull(force)?;
        // only build if the environment was updated
        if let PullResult::Updated = state {
            env.build(flox)?;
        }
        Ok(state)
```

File: cli/flox/src/commands/environment.rs [hunk 3]
```diff
@@ -1627,17 +1628,24 @@ impl Pull {
...
+                            You can activate this environment with 'flox activate'
+                            ",
+                            owner = pointer.owner, name = pointer.name,
+                            floxhub_host = flox.floxhub.base_url(),
+                            suffix = if self.force { " (forced)" } else { "" }
+                        });
+                    },
+                    PullResult::UpToDate => {
+                        message::warning(formatdoc! {"
+                            {owner}/{name} is already up to date.
```
Comment (mkenigs) [inline cli/flox/src/commands/environment.rs:1645]: nit/question: do you want the trailing newline?

**Review by mkenigs — APPROVED** [review]

File: cli/flox/src/commands/environment.rs (hunk 2 above)
Comment (ysndr) [reply cli/flox/src/commands/environment.rs:1673]: Agree, applied

File: cli/flox/src/commands/environment.rs (hunk 3 above)
Comment (ysndr) [reply cli/flox/src/commands/environment.rs:1645]: It's a bit inconsistent that our `formatdoc!` messages end with a newline and `format!` ones don't, I agree
It's due to how we write them in the source, i.e. `"` on a separate line, so we could change those treewide. but I'd rather do that separately from this PR

**Comment by ghudgins** [comment]
I tried this on the branch by pulling twice and got 

"❌ ERROR: Cannot pull a new environment into an existing one"

image

**Comment by ysndr** [comment]
You ran `flox pull <remote>` which tries to pull a _new_ env, try just `flox pull`

**Review by mkenigs — APPROVED** [review]

**Pull request merged**